torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+        we remove them and try to reconstruct them in :func:`__setstate__`.
+        """
+        state = copy.copy(self.__dict__)
+        state["is_sharded"] = [p._is_sharded for p in self.params]  # type: ignore[attr-defined]
+        state["orig_sizes"] = [p._orig_size for p in self.params]  # type: ignore[attr-defined]
+        if state["process_group"] is not None:
+            state["process_group"] = "MISSING"  # process_group isn't pickleable
+        self._reset_lazy_init()
+        return state
+
+    def __setstate__(self, state: Dict[str, Any]) -> None:
+        """Intercept state setting and perform needed changes on params."""
+        super().__setstate__(state)
+
+        def fixup(p: Parameter, is_sharded: bool, size: torch.Size) -> Parameter:
+            assert isinstance(p, Parameter)
+            p.data = p.data.clone()  # move tensors out of shared memory
+            p._is_sharded = is_sharded  # type: ignore[attr-defined]
+            p._orig_size = size  # type: ignore[attr-defined]
+            return p
+
+        self.params = [
+            fixup(p, is_sharded, size) for p, is_sharded, size in zip(self.params, self.is_sharded, self.orig_sizes)
+        ]
+        del self.is_sharded
+        del self.orig_sizes
+        self._reset_lazy_init()
+
+    def __getitem__(self, key: int) -> Any:
+        """Forward indexing calls in case the module is a nn.Sequential."""
+        return self.module.__getitem__(key)  # type: ignore[operator]
+
+    def _reset_lazy_init(self) -> None:
+        """Reset instance so :func:`_lazy_init` will run on the next forward."""
+        self._is_root: Optional[bool] = None
+        self._streams: Dict[str, torch.cuda.Stream] = {}
+        for p in self.params:
+            if hasattr(p, "_local_shard"):
+                # reset _init_param_attributes
+                del p._local_shard  # type: ignore[attr-defined]
+
+    def _lazy_init(self) -> None:
+        """Initialization steps that should happen lazily, typically right
+        before the first forward pass.
+        """
+        # Initialize param attributes lazily, in case the param's dtype or
+        # device changes after __init__.
+        for p in self.params:
+            self._init_param_attributes(p)
+
+        # Initialize _is_root and setup streams. These steps would ideally
+        # happen in __init__, but _is_root can only be determined after the
+        # entire model hierarchy is setup, thus we run it lazily.
+        if self._is_root is None:
+            self._set_is_root()
+            self._setup_streams()
+
+        if self._is_root:
+            # Buffers stay on GPU, and don't get sharded. Since _cast_buffers
+            # applies recursively, we only call this from the root instance.
+            self._cast_buffers()
+
+            # Don't free the full params for the outer-most (root) instance,
+            # since those params will be needed immediately after for the
+            # backward pass.
```

Comment:
Is this statement always true? IIUC, the outer-most FSDP instance covers the remaining parameters that are not wrapped by FSDP, which might not be the last layer? E.g., I assume in the following case, the outer-most FSDP covers the first layer instead of the last layer?

```python
net = FSDP(nn.Sequential([
  nn.Linear(2, 2),
  FSDP(nn.Linear(2, 2)),
]))
```

No need to change this in the first PR, but if the above assessment if true, let's create an issue to track

Reply:
your understanding is correct, it does not have to be last layers. it assumes most of cases outer most instance will either not own any params or own last layers, so adding this optimization in default